src/pick.c

```diff
@@ -608,62 +643,70 @@ put_line(char *string, int length)
 void
 print_query(char *query, size_t length, size_t position, size_t scroll)
 {
-	size_t	i;
+	size_t	i = 0;
 
 	tty_putp(restore_cursor);
 	put_line(query + scroll, length - scroll);
 
 	tty_putp(restore_cursor);
-	for (i = 0; i < position - scroll; ++i)
+	while (i < position - scroll) {
+		while (isu8cont(query[++i]));
 		tty_putp(cursor_right);
+	}
 }
 
 int
 print_choices(int selection)
 {
-	int		 i, j, k;
-	size_t		 length, query_length;
+	int		 col, i, j, k;
+	size_t		 query_length;
 	struct choice	*choice;
 
 	/* Emit query line. */
 	tty_putc('\n');
 
 	query_length = strlen(query);
-	for (choice = choices.v, i = 0;
+	for (choice = choices.v, col = i = 0;
 	     i < (ssize_t)choices.length
 	     && i < lines - 1
 	     && (query_length == 0 || choice->score > 0);
 	     choice++, i++) {
 		if (i == selection)
 			tty_putp(enter_standout_mode);
 
-		for (j = 0, length = MIN(choice->match_start, (size_t)columns);
-		     j < (ssize_t)length;
-		     j++)
+		for (col = j = 0;
+		     j < (ssize_t)choice->match_start && col < columns;
+		     col += !isu8cont(choice->string[j]), j++)
 			if (tty_putc(choice->string[j]) == EOF)
 				err(1, "tty_putc");
 
 		tty_putp(enter_underline_mode);
 
-		for (length = MIN(choice->match_end, (size_t)columns);
-		     j < (ssize_t)length;
-		     j++)
+		for (;
+		     j < (ssize_t)choice->match_end && col < columns;
+		     col += !isu8cont(choice->string[j]), j++)
 			if (tty_putc(choice->string[j]) == EOF)
 				err(1, "tty_putc");
 
 		tty_putp(exit_underline_mode);
 
-		for (length = MIN(choice->length, (size_t)columns);
-		     j < (ssize_t)length;
-		     j++)
+		for (;
+		     j < (ssize_t)choice->length && col < columns;
+		     col += !isu8cont(choice->string[j]), j++) {
 			/* A null character will be present before the
 			 * terminating null character if descriptions is
 			 * enabled. */
-			if (tty_putc(choice->string[j] == '\0'
-				     ? ' ' : choice->string[j]) == EOF)
-				err(1, "tty_putc");
+			if (choice->string[j] == '\0') {
+				if (tty_putc(' ') == EOF)
+					err(1, "tty_putc");
+			} else {
+				if (tty_putc(choice->string[j]) == EOF)
```

Comment:
On Fri, Apr 08, 2016 at 01:43:14AM -0700, Calle Erlandsson wrote:
> >  			/* A null character will be present before the
> >  			 * terminating null character if descriptions is
> >  			 * enabled. */
> > -			if (tty_putc(choice->string[j] == '\0'
> > -				     ? ' ' : choice->string[j]) == EOF)
> > -				err(1, "tty_putc");
> > +			if (choice->string[j] == '\0') {
> > +				if (tty_putc(' ') == EOF)
> > +					err(1, "tty_putc");
> > +			} else {
> > +				if (tty_putc(choice->string[j]) == EOF)
> 
> Could this `if` be joined with the above `else` to reduce indentation?

Nice catch, fixed in 727b8f6.